# Development Tutorial
## Getting Started
This tutorial focuses on selecting the development factors. 

Be sure to make sure your packages are updated. For more info on how to update your pakages, visit [Keeping Packages Updated](https://chainladder-python.readthedocs.io/en/latest/library/install.html#keeping-packages-updated).

In [1]:
# Black linter, optional
%load_ext lab_black

import pandas as pd
import numpy as np
import chainladder as cl

print("pandas: " + pd.__version__)
print("numpy: " + np.__version__)
print("chainladder: " + cl.__version__)

pandas: 1.4.2
numpy: 1.22.4
chainladder: 0.8.12


## Disclaimer
Note that a lot of the examples shown might not be applicable in a real world scenario, and is only meant to demonstrate some of the functionalities included in the package. The user should always follow all applicable laws, the Code of Professional Conduct, applicable Actuarial Standards of Practice, and exercise their best actuarial judgement.

## Testing for Violation of Chain Ladder's Assumptions

The chain ladder method is based on the strong assumptions of independence across origin periods and across valuation periods. Mack developed tests to verify if these assumptions hold, and these tests have been implemented in the `chainladder` package.

Before the chain ladder model can be used, we should verify that the data satisfies the underlying assumptions using tests at the desired confidence interval level. If assumptions are violated, we should consider if ultimates can be estimated using other models.

There are two main tests that we need to perform:
- The `valuation_correlation` test: 
    - This test tests for the assumption of independence of accident years. In fact, it tests for correlation across calendar periods (diagonals), and by extension, origin periods (rows).
    - An additional parameter, `total`, can be passed, depending on if we want to calculate valuation correlation in total across all origins (`True`), or for each origin separately (`False`).
    - The test uses Z-statistic.
- The `development_correlation` test:
    - This test tests for the assumption of independence of the chain ladder method that assumes that subsequent development factors are not correlated (columns).
    - The test uses T-statistic.

In [2]:
raa = cl.load_sample("raa")
print(
    "Are valuation years correlated? Or, are the origins correlated?",
    raa.valuation_correlation(p_critical=0.1, total=True).z_critical.values,
)
print(
    "Are development periods coorelated?",
    raa.development_correlation(p_critical=0.5).t_critical.values,
)

Are valuation years correlated? Or, are the origins correlated? [[False]]
Are development periods coorelated? [[False]]


The above tests show that the `raa` triangle is independent in both cases, suggesting that there is no evidence that the chain ladder model is not an appropriate method to develop the ultimate amounts. It is suggested to review Mack's papers to ensure a proper understanding of the methodology and the choice of `p_critical`.

Mack also demonstrated that we can test for valuation years' correlation. To test for each valuation year's correlation individually, we set `total` to `False`.

In [3]:
raa.valuation_correlation(p_critical=0.1, total=False).z_critical

1982   1983   1984   1985   1986   1987   1988   1989   1990
1981  False  False  False  False  False  False  False  False  False

Note that the tests are run on the entire 4 dimensions of the `triangle`.

## Estimator Basics

All development methods follow the `sklearn` estimator API.  These estimators have a few properties that are worth getting used to.

We instantiate the estimator with your choice of assumptions.  In the case where we don't opt for any assumptions, defaults are chosen for you.

At this point, we've chosen an estimator and assumptions (even if default) but we have not shown our estimator a `Triangle`.  At this point it is merely instructions on how to fit development patterns, but no patterns exist as of yet.

All estimators have a `fit` method and you can pass a triangle to your estimator.  Let's `fit` a `Triangle` in a `Development` estimator.  Let's also assign the estimator to a variable so we can reference attributes about it.

In [4]:
genins = cl.load_sample("genins")
dev = cl.Development().fit(genins)

Now that we have `fit` a `Development` estimator, it has many additional properties that didn't exist before fitting.  For example, 
we can view the `ldf_`

In [5]:
dev.ldf_

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  3.490607  1.747333  1.457413  1.173852  1.103824  1.086269  1.053874  1.076555  1.017725

We can view the `cdf_`

In [6]:
dev.cdf_

12-Ult    24-Ult    36-Ult    48-Ult    60-Ult    72-Ult    84-Ult    96-Ult   108-Ult
(All)  14.446577  4.138701  2.368582  1.625196  1.384499  1.254276  1.154664  1.095637  1.017725

We can also convert between LDFs and CDFs using incr_to_cum() and cum_to_incr() similar to triangles.

In [7]:
dev.ldf_.incr_to_cum()

12-Ult    24-Ult    36-Ult    48-Ult    60-Ult    72-Ult    84-Ult    96-Ult   108-Ult
(All)  14.446577  4.138701  2.368582  1.625196  1.384499  1.254276  1.154664  1.095637  1.017725

In [8]:
dev.cdf_.cum_to_incr()

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  3.490607  1.747333  1.457413  1.173852  1.103824  1.086269  1.053874  1.076555  1.017725

Notice these attributes have a trailing underscore (`_`). This is scikit-learn's API convention, as its [documentation](https://scikit-learn.org/dev/developers/develop.html) states, "attributes that have been estimated from the data must always have a name ending with trailing underscore, for example the coefficients of some regression estimator would be stored in a `coef_` attribute after `fit` has been called." In summary, the trailing underscore in class attributes is a scikit-learn's convention to denote that the attributes are estimated, or to denote that they are fitted attributes.

In [9]:
print("Assumption parameter (no underscore):", dev.average)
print("Estimated parameter (underscore):\n", dev.ldf_)

Assumption parameter (no underscore): volume
Estimated parameter (underscore):
           12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  3.490607  1.747333  1.457413  1.173852  1.103824  1.086269  1.053874  1.076555  1.017725


## Development Averaging

Now that we have a grounding in triangle manipulation and the basics of estimators, we can start getting more creative with customizing our development factors.

The basic `Development` estimator uses a weighted regression through the origin for estimating parameters. Mack showed that using weighted regressions allows for:
1. `volume` weighted average development patterns<br>
2. `simple` average development factors<br>
3. OLS `regression` estimate of development factor where the regression equation is Y = mX + 0<br>

While he posited this framework to suggest the `MackChainladder` stochastic method, it is an elegant form even for deterministic development pattern selection.

In [10]:
genins = cl.load_sample("genins")
genins

12         24         36         48         60         72         84         96         108        120
2001  357848.0  1124788.0  1735330.0  2218270.0  2745596.0  3319994.0  3466336.0  3606286.0  3833515.0  3901463.0
2002  352118.0  1236139.0  2170033.0  3353322.0  3799067.0  4120063.0  4647867.0  4914039.0  5339085.0        NaN
2003  290507.0  1292306.0  2218525.0  3235179.0  3985995.0  4132918.0  4628910.0  4909315.0        NaN        NaN
2004  310608.0  1418858.0  2195047.0  3757447.0  4029929.0  4381982.0  4588268.0        NaN        NaN        NaN
2005  443160.0  1136350.0  2128333.0  2897821.0  3402672.0  3873311.0        NaN        NaN        NaN        NaN
2006  396132.0  1333217.0  2180715.0  2985752.0  3691712.0        NaN        NaN        NaN        NaN        NaN
2007  440832.0  1288463.0  2419861.0  3483130.0        NaN        NaN        NaN        NaN        NaN        NaN
2008  359480.0  1421128.0  2864498.0        NaN        NaN        NaN        NaN        NaN        NaN        NaN
2009  376686.0  1363294.0        NaN        NaN        NaN        NaN        NaN        NaN        NaN        NaN
2010  344014.0        NaN        NaN        NaN        NaN        NaN        NaN        NaN        NaN        NaN

We can also print the `age_to_age` factors.

In [11]:
genins.age_to_age

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
2001  3.143200  1.542806  1.278299  1.237719  1.209207  1.044079  1.040374  1.063009  1.017725
2002  3.510582  1.755493  1.545286  1.132926  1.084493  1.128106  1.057268  1.086496       NaN
2003  4.448450  1.716718  1.458257  1.232079  1.036860  1.120010  1.060577       NaN       NaN
2004  4.568002  1.547052  1.711784  1.072518  1.087360  1.047076       NaN       NaN       NaN
2005  2.564198  1.872956  1.361545  1.174217  1.138315       NaN       NaN       NaN       NaN
2006  3.365588  1.635679  1.369162  1.236443       NaN       NaN       NaN       NaN       NaN
2007  2.922798  1.878099  1.439393       NaN       NaN       NaN       NaN       NaN       NaN
2008  3.953288  2.015651       NaN       NaN       NaN       NaN       NaN       NaN       NaN
2009  3.619179       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN

And colorcode with `heatmap()`.

In [12]:
genins.age_to_age.heatmap()

/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/chainladder/core/display.py:134: FutureWarning: this method is deprecated in favour of `Styler.to_html()`
  .render()


In [13]:
vol = cl.Development(average="volume").fit(genins).ldf_
vol

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  3.490607  1.747333  1.457413  1.173852  1.103824  1.086269  1.053874  1.076555  1.017725

In [14]:
sim = cl.Development(average="simple").fit(genins).ldf_
sim

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  3.566143  1.745557  1.451961  1.180984  1.111247  1.084818  1.052739  1.074753  1.017725

In most cases, estimator attributes are `Triangle`s themselves and can be manipulated with just like raw triangles.

In [15]:
print("LDF Type: ", type(vol))
print("Difference between volume and simple average:")
vol - sim

LDF Type:  <class 'chainladder.core.triangle.Triangle'>
Difference between volume and simple average:


12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108  108-120
(All) -0.075536  0.001776  0.005452 -0.007132 -0.007423  0.001452  0.001135  0.001802      NaN

We can specify how the LDFs are averaged independently for each age-to-age period. For example, we can use `volume` averaging on the first pattern, `simple` the second, `regression` the third, and then repeat the cycle three times for the 9 age-to-age factors that we need. Note that the array of selected method must be of the same length as the number of age-to-age factors.

In [16]:
cl.Development(average=["volume", "simple", "regression"] * 3).fit(genins).ldf_

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  3.490607  1.745557  1.461852  1.173852  1.111247  1.087341  1.053874  1.074753  1.017725

Another example, using `volume`-weighting for the first factor, `simple`-weighting for the next 5 factors, and `volume`-weighting for the last 3 factors.

In [17]:
cl.Development(average=["volume"] + ["simple"] * 5 + ["volume"] * 3).fit(genins).ldf_

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  3.490607  1.745557  1.451961  1.180984  1.111247  1.084818  1.053874  1.076555  1.017725

## Averaging Period

`Development` comes with an `n_periods` parameter that allows you to select the latest `n` origin periods for fitting your development patterns. `n_periods=-1` is used to indicate the usage of all available periods, which is also the default if the parameter is not specified. The units of `n_periods` follows the `origin_grain` of the underlying triangle.

In [18]:
cl.Development().fit(genins).ldf_

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  3.490607  1.747333  1.457413  1.173852  1.103824  1.086269  1.053874  1.076555  1.017725

In [19]:
cl.Development(n_periods=-1).fit(genins).ldf_

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  3.490607  1.747333  1.457413  1.173852  1.103824  1.086269  1.053874  1.076555  1.017725

In [20]:
cl.Development(n_periods=3).fit(genins).ldf_

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  3.460401  1.846507  1.392009  1.153852  1.084915  1.097355  1.053874  1.076555  1.017725

Much like `average`, `n_periods` can also be set for each age-to-age individually.

In [21]:
cl.Development(n_periods=[8, 2, 6, 5, -1, 2, -1, -1, 5]).fit(genins).ldf_

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  3.532471  1.950242  1.480761  1.165122  1.103824  1.082476  1.053874  1.076555  1.017725

Note that if we provide `n_periods` that is greater than what is available for any particular age-to-age period, all available periods will be used instead.

In [22]:
cl.Development(n_periods=[1, 2, 3, 4, 5, 6, 7, 8, 9]).fit(
    genins
).ldf_ == cl.Development(n_periods=[1, 2, 3, 4, 5, 4, 3, 2, 1]).fit(genins).ldf_

True

## Discarding Problematic Link Ratios

Even with `n_periods`, there are situations where you might want to be more surgical in our selections. For example, you could have a valuation period with bad data and wish to omit the entire diagonal from your averaging.

In [23]:
cl.Development(drop_valuation="2004").fit(genins).ldf_

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  3.379677  1.751749  1.442605  1.165122  1.103824  1.086269  1.053874  1.076555  1.017725

We can also do an olympic averaging (i.e. exluding high and low from each period).

In [24]:
cl.Development(drop_high=True, drop_low=True).fit(genins).ldf_

/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/chainladder/development/base.py:144: UserWarning: Some exclusions have been ignored. At least 1 (use preserve = ...) link ratio(s) is required for development estimation.
  warnings.warn(warning)


12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  3.520098  1.727701  1.435147  1.193021  1.101827  1.082476  1.057268  1.076555  1.017725

The function also accepts intergers. For example, if we want to drop the highest 3 factors from each period.

In [25]:
cl.Development(drop_high=3).fit(genins).ldf_

/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/chainladder/development/base.py:144: UserWarning: Some exclusions have been ignored. At least 1 (use preserve = ...) link ratio(s) is required for development estimation.
  warnings.warn(warning)


12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  3.161368  1.639211  1.368696  1.122203  1.060105  1.044079  1.053874  1.076555  1.017725

There's a `preserve` that we can use, this variable allows us to specified the minimum number of LDFs required for calculation. If this minimum is not yet, the `drop_high` and `drop_low` for that age will be ignored. This is especially useful in the tail when the data is thin.

In [26]:
cl.Development(drop_high=3, drop_low=2, preserve=2).fit(genins).ldf_

/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/chainladder/development/base.py:144: UserWarning: Some exclusions have been ignored. At least 2 link ratio(s) is required for development estimation.
  warnings.warn(warning)


12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  3.410772  1.701152  1.406103  1.173852  1.103824  1.086269  1.053874  1.076555  1.017725

We can also use an array of booleans or ints.

In [27]:
cl.Development(drop_high=[True, True, False, True], drop_low=[1, 2, 0, 3]).fit(
    genins
).ldf_

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  3.520098  1.768474  1.457413  1.234173  1.103824  1.086269  1.053874  1.076555  1.017725

Or maybe there is just a single outlier link-ratio that you don't think is indicative of future development.  For these, you can specify the intersection of the origin and development age of the **denominator** of the link-ratio to `drop`.

In [28]:
genins.age_to_age.heatmap()

/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/chainladder/core/display.py:134: FutureWarning: this method is deprecated in favour of `Styler.to_html()`
  .render()


Let's say we believe the 4.5680 factor from origin 2004 between age 12 and 24 should be dropped, we can use `drop=('2004', 12)`.

In [29]:
cl.Development(drop=("2004", 12)).fit(genins).ldf_

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  3.379677  1.747333  1.457413  1.173852  1.103824  1.086269  1.053874  1.076555  1.017725

If there are more than one outliers, you can also pass an array of array to the `drop` argument.

In [30]:
cl.Development(drop=[("2004", 12), ("2008", 24)]).fit(genins).ldf_

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  3.379677  1.704149  1.457413  1.173852  1.103824  1.086269  1.053874  1.076555  1.017725

## Transformers
In `sklearn`, there are two types of estimators: transformers and predictors. A transformer transforms the input data (X) in some ways, and a predictor predicts a new value (or values, Y) by using the input data X.

`Development` is a transformer, as the returned object is a means to create development patterns, which is used to estimate ultimates, but itself is not a reserving model (predictor). 

Transformers come with the `tranform` and `fit_transform` method. These will return a `Triangle` object, but augment it with additional information for use in a subsequent IBNR model (a predictor). `drop_high` (and `drop_low`) can take an array of boolean variables, indicating if the highest factor should be dropped for each of the LDF calculation.

In [31]:
transformed_triangle = cl.Development(drop_high=[True] * 4 + [False] * 5).fit_transform(
    genins
)
transformed_triangle

12         24         36         48         60         72         84         96         108        120
2001  357848.0  1124788.0  1735330.0  2218270.0  2745596.0  3319994.0  3466336.0  3606286.0  3833515.0  3901463.0
2002  352118.0  1236139.0  2170033.0  3353322.0  3799067.0  4120063.0  4647867.0  4914039.0  5339085.0        NaN
2003  290507.0  1292306.0  2218525.0  3235179.0  3985995.0  4132918.0  4628910.0  4909315.0        NaN        NaN
2004  310608.0  1418858.0  2195047.0  3757447.0  4029929.0  4381982.0  4588268.0        NaN        NaN        NaN
2005  443160.0  1136350.0  2128333.0  2897821.0  3402672.0  3873311.0        NaN        NaN        NaN        NaN
2006  396132.0  1333217.0  2180715.0  2985752.0  3691712.0        NaN        NaN        NaN        NaN        NaN
2007  440832.0  1288463.0  2419861.0  3483130.0        NaN        NaN        NaN        NaN        NaN        NaN
2008  359480.0  1421128.0  2864498.0        NaN        NaN        NaN        NaN        NaN        NaN        NaN
2009  376686.0  1363294.0        NaN        NaN        NaN        NaN        NaN        NaN        NaN        NaN
2010  344014.0        NaN        NaN        NaN        NaN        NaN        NaN        NaN        NaN        NaN

In [32]:
transformed_triangle.link_ratio

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
2001  3.143200  1.542806  1.278299       NaN  1.209207  1.044079  1.040374  1.063009  1.017725
2002  3.510582  1.755493  1.545286  1.132926  1.084493  1.128106  1.057268  1.086496       NaN
2003  4.448450  1.716718  1.458257  1.232079  1.036860  1.120010  1.060577       NaN       NaN
2004       NaN  1.547052       NaN  1.072518  1.087360  1.047076       NaN       NaN       NaN
2005  2.564198  1.872956  1.361545  1.174217  1.138315       NaN       NaN       NaN       NaN
2006  3.365588  1.635679  1.369162  1.236443       NaN       NaN       NaN       NaN       NaN
2007  2.922798  1.878099  1.439393       NaN       NaN       NaN       NaN       NaN       NaN
2008  3.953288       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN
2009  3.619179       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN

Our transformed triangle behaves as our original `genins` triangle.  However, notice the link_ratios exclude any droppped values you specified.

In [33]:
transformed_triangle.link_ratio.heatmap()

/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/chainladder/core/display.py:134: FutureWarning: this method is deprecated in favour of `Styler.to_html()`
  .render()


In [34]:
print(type(transformed_triangle))
transformed_triangle.latest_diagonal

<class 'chainladder.core.triangle.Triangle'>


2010
2001  3901463.0
2002  5339085.0
2003  4909315.0
2004  4588268.0
2005  3873311.0
2006  3691712.0
2007  3483130.0
2008  2864498.0
2009  1363294.0
2010   344014.0

However, it has other attributes that make it IBNR model-ready.

In [35]:
transformed_triangle.cdf_

12-Ult    24-Ult   36-Ult   48-Ult    60-Ult    72-Ult    84-Ult    96-Ult   108-Ult
(All)  13.136729  3.886978  2.28089  1.61311  1.384499  1.254276  1.154664  1.095637  1.017725

`fit_transform()` is equivalent to calling `fit` and `transform` in succession on the same triangle.  Again, this should feel very familiar to the `sklearn` practitioner.

In [36]:
cl.Development().fit_transform(genins) == cl.Development().fit(genins).transform(genins)

True

The reason you might want want to use `fit` and `transform` separately would be when you want to apply development patterns to a a different triangle.  For example, we can:

1. Extract the commercial auto triangles from the `clrd` dataset<br>
2. Summarize to an industry level and `fit` a `Development` object<br>
3. We can then `transform` the individual company triangles with the industry development patterns<br>

In [37]:
clrd = cl.load_sample("clrd")
comauto = clrd[clrd["LOB"] == "comauto"]["CumPaidLoss"]

comauto_industry = comauto.sum()
industry_dev = cl.Development().fit(comauto_industry)

industry_dev.transform(comauto)

Triangle Summary
Valuation:           1997-12
Grain:                  OYDY
Shape:      (157, 1, 10, 10)
Index:         [GRNAME, LOB]
Columns:       [CumPaidLoss]

## Working with Multidimensional Triangles

Several (though not all) of the estimators in `chainladder` can be fit to several triangles simultaneously. While this can be a convenient shorthand, all these estimators use the same assumptions across every triangle.

In [38]:
clrd = cl.load_sample("clrd").groupby("LOB").sum()["CumPaidLoss"]
print("Fitting to " + str(len(clrd.index)) + " industries simultaneously.")
cl.Development().fit_transform(clrd).cdf_

Fitting to 6 industries simultaneously.


Triangle Summary
Valuation:          2261-12
Grain:                 OYDY
Shape:         (6, 1, 1, 9)
Index:                [LOB]
Columns:      [CumPaidLoss]

For greater control, you can slice individual triangles out and fit separate patterns to each.

In [39]:
print(cl.Development(average="simple").fit(clrd.loc["wkcomp"]))
print(cl.Development(n_periods=4).fit(clrd.loc["ppauto"]))
print(cl.Development(average="regression", n_periods=6).fit(clrd.loc["comauto"]))

Development(average='simple')
Development(n_periods=4)
Development(average='regression', n_periods=6)
